bundles/org.openhab.binding.meater/src/main/java/org/openhab/binding/meater/internal/api/MeaterRestAPI.java

```diff
@@ -179,13 +188,14 @@ private String getFromApi(String uri) throws MeaterException, MeaterAuthenticati
         }
     }
 
-    public <T> T getDevices(Class<T> dto) throws MeaterException, MeaterAuthenticationException {
+    public <T> T getDevices(Class<T> dto) throws MeaterException {
         String uri = DEVICES;
         String json = "";
 
         try {
             json = getFromApi(uri);
-        } catch (MeaterAuthenticationException e) {
+        } catch (MeaterException e) {
```

Comment:
This change is wrong, you need to catch `MeaterAuthenticationException` explicitly. Other MeaterExceptions should not cause authentication reattempt.

Reply:
True, fixed!

Reply:
Ah, now I think I understand why. `login()` itself doesn't throw MeaterAuthenticationException, only `getFromApi()` does when receiving 401. And `getFromApi()` is still responsible for the initial authentication. I guess question is what to do if this fails?

Reply:
It shouldn't try to reauthenticate for errors unrelated to missing authentication. IMHO the cleanest approach would be to also get rid of:

```java
                    if (authToken.isEmpty()) {
                        login();
                    }
```

from `getFromApi()` and move it to `getDevices()` so `getFromApi()` wouldn't have any responsibility connected to authentication except throwing an exception for 401 so reauthentication logic can be handled by the caller.

Looking at `login()` again I think it's also okay as it is. Of course it could retry on timeout, but it doesn't seem important, as it will also be retried on next cycle, and probably it wouldn't help much if the API would start to cause timeouts.

Reply:
Like this then?

```
    public @Nullable <T> T getDevices(Class<T> dto) throws MeaterException {
        String uri = DEVICES;
        String json = "";

        if (authToken.isEmpty()) {
            login();
        }

        try {
            json = getFromApi(uri);
        } catch (MeaterAuthenticationException e) {
            logger.debug("getFromApi failed {}", e.getMessage());
            this.authToken = "";
            login();
            json = getFromApi(uri);
        }
```